src/StableHashTraits.jl

```diff
@@ -392,22 +414,101 @@ end
 #####
 
 qname_(T, name) = validate_name(cleanup_name(string(parentmodule(T), '.', name(T))))
-qualified_name(fn::Function) = qname_(fn, nameof)
-qualified_type(fn::Function) = qname_(fn, string)
-qualified_name(x::T) where {T} = qname_(T <: DataType ? x : T, nameof)
-qualified_type(x::T) where {T} = qname_(T <: DataType ? x : T, string)
+qualified_name_(fn::Function) = qname_(fn, nameof)
+qualified_type_(fn::Function) = qname_(fn, string)
+qualified_name_(x::T) where {T} = qname_(T <: DataType ? x : T, nameof)
+qualified_type_(x::T) where {T} = qname_(T <: DataType ? x : T, string)
+function qualified_name(x)
+    Base.depwarn("`qualified_name` is deprecated, favor `stable_typename_id` in all cases " *
+                 "where backwards compatible hash values are not required.",
+                 :qualified_name)
+    return qualified_name_(x)
+end
+function qualified_type(x)
+    Base.depwarn("`qualified_type` is deprecated, favor `stable_type_id` in all cases " *
+                 "where backwards compatible hash values are not required.",
+                 :qualified_type)
+    return qualified_type_(x)
+end
+
+function hash_type_str(str, T)
+    bytes = sha256(codeunits(str))
+    return first(reinterpret(UInt64, bytes))
```

Comment:
I've tried to add an explanation about the goal in a comment. 

To explain with perhaps excessive thoroughness: We want 64 bytes, because the objective is to have a relatively fast-to-hash representation of each type, and 256 bytes seems excessive. But also, we want as low a probability of hash collisions as I can reasonably achieve in those 64 bytes (since we're computing this at compile time in any case). There's no sha64, so I just take the smallest sha available, and take the bytes of that hash that I need.

Reply:
ok, but aren't you computing those 256 bytes anyway and just throwing most of them away? or is it that you compute _another hash_ on those 64 bytes which is signifcantly faster than?

Reply:
Ah okay! I think see the confusion; maybe it is worth spelling out what a generated function is.

A generated function is effectively a special kind of macro. It gets when a given concrete method is being compiled. It is passed the *types* of the arguments and is expected to return an expression. This expression is then compiled as the source code of that specific method.

So, when I compute the `sha256` hash inside the generated function it is only evaluated once, when the method gets compiled. The code that actually gets compiled looks something like `:(return UInt(1))`. On future calls to said function no hash is computed, it just returns a number. 

I want that number to be small-ish since it can end up be a significant proportion of the bytes that get hashed for the worst case scenarios, e.g. arrays of small structs.